pkg/controller/operatorsource/operatorsource_controller.go

```diff
@@ -0,0 +1,140 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+/**
+* USER ACTION REQUIRED: This is a scaffold file intended for the user to modify with their own Controller
+* business logic.  Delete these comments after modifying this file.*
+ */
+
+// Add creates a new OperatorSource Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	client := mgr.GetClient()
+	scheme := mgr.GetScheme()
+
+	phaseReconcilerFactory := &phaseReconcilerFactory{
+		registryClientFactory: appregistry.NewClientFactory(),
+		datastore:             datastore.Cache,
+		client:                client,
+	}
+
+	transitioner := phase.NewTransitioner()
+
+	return &ReconcileOperatorSource{Client: client, Scheme: scheme, Factory: phaseReconcilerFactory, Transitioner: transitioner, OperatorSourceGenerator: &DefaultOperatorSourceGenerator{}}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("operatorsource-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource OperatorSource
+	err = c.Watch(&source.Kind{Type: &marketplacev1alpha1.OperatorSource{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileOperatorSource{}
+
+// ReconcileOperatorSource reconciles a OperatorSource object
+type ReconcileOperatorSource struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	Client                  client.Client
+	Scheme                  *runtime.Scheme
+	Factory                 PhaseReconcilerFactory
+	Transitioner            phase.Transitioner
+	OperatorSourceGenerator OperatorSourceGenerator
+}
+
+// Reconcile reads that state of the cluster for a OperatorSource object and makes changes based on the state read
+// and what is in the OperatorSource.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileOperatorSource) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Printf("Reconciling OperatorSource %s/%s\n", request.Namespace, request.Name)
+
+	// Fetch the OperatorSource instance
+	//instance := &marketplacev1alpha1.OperatorSource{}
+	instance := r.OperatorSourceGenerator.GenerateOperatorSource()
```

Comment:
I would like to know more about why it is needed.

Reply:
The client.GET() method takes the address of an empty resource that has already been allocated and fills the struct with the necessary data. Because of that pattern, it isn't possible for us to just mock the GET() call. The only way to deal with this is either modify the reconcile function to somehow get an OperatorSource that has already been allocated with the data we need for comparison, or to create an instance of a struct that generates the object with the right data upfront and mock that.

Reply:
Please add a comment about this.

Reply:
If we leave `pkg/operatorsource` as is, then it becomes a concern for `pkg/controller/operatorsource` package. Maybe FakeClient handles this, I will look for more details.

Reply:
This was removed for now, we can address unit testing in a separate story